### PR TITLE
uadk: support openssl 3.0

### PR DESCRIPTION
**Pull request by gaozhangfei**
Openssl3.0 provoder will be built when libcrypto >= 3.0

First step to support digest

Test with
openssl speed -provider xx/uadk_provider.so -provider default -evp md5 openssl speed -provider xx/uadk_provider.so -provider default -evp sm3 openssl speed -provider xx/uadk_provider.so -provider default -evp sha1 openssl speed -provider xx/uadk_provider.so -provider default -evp sha3-224 openssl speed -provider xx/uadk_provider.so -provider default -evp sha3-256 openssl speed -provider xx/uadk_provider.so -provider default -evp sha3-384 openssl speed -provider xx/uadk_provider.so -provider default -evp sha3-512

Signed-off-by: Zhangfei Gao <zhangfei.gao@linaro.org>

### CODE REVIEW

**Comment by CaptainMiaow** [comment]
Does "e_uadk.c" need to be renamed to "uadk_engine_init.c" ?

File: src/uadk_prov_init.c [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+
+static OSSL_FUNC_provider_query_operation_fn p_prov_query;
+static OSSL_FUNC_provider_teardown_fn p_teardown;
+
+static const OSSL_ALGORITHM *p_prov_query(void *provctx, int operation_id,
+					  int *no_cache)
+{
+	*no_cache = 0;
+	switch (operation_id) {
```
Comment (CaptainMiaow) [inline src/uadk_prov_init.c:54]: Maybe a blank line is needed.

File: src/uadk_prov_init.c [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+};
+
+
+static OSSL_FUNC_provider_query_operation_fn p_prov_query;
+static OSSL_FUNC_provider_teardown_fn p_teardown;
+
+static const OSSL_ALGORITHM *p_prov_query(void *provctx, int operation_id,
+					  int *no_cache)
+{
+	*no_cache = 0;
```
Comment (CaptainMiaow) [inline src/uadk_prov_init.c:53]: Maybe a blank line is needed after line53.

File: src/uadk_prov_init.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		       const OSSL_DISPATCH *oin,
+		       const OSSL_DISPATCH **out,
+		       void **provctx)
+{
+	struct p_uadk_ctx *ctx;
+	const OSSL_DISPATCH *in = oin;
+
+	for (; in->function_id != 0; in++) {
+		switch (in->function_id) {
+		}
```
Comment (CaptainMiaow) [inline src/uadk_prov_init.c:77]: There is an empty switch() branch.

File: src/uadk_prov_init.c [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+int OSSL_provider_init(const OSSL_CORE_HANDLE *handle,
+		       const OSSL_DISPATCH *oin,
+		       const OSSL_DISPATCH **out,
+		       void **provctx)
+{
+	struct p_uadk_ctx *ctx;
+	const OSSL_DISPATCH *in = oin;
+
+	for (; in->function_id != 0; in++) {
```
Comment (CaptainMiaow) [inline src/uadk_prov_init.c:75]: Do we need to determine whether parameter "handle", "oin", "out" and "provctx"are empty before using in this _init() function?

File: src/uadk_digest.c [hunk 5]
```diff
@@ -30,6 +30,11 @@
 #include "uadk_async.h"
```
Comment (CaptainMiaow) [inline src/uadk_digest.c:30]: Do we need to use a separate .c file to store the provider implementation of each algorithm?

File: src/uadk.h [hunk 6]
```diff
@@ -40,10 +40,22 @@ void uadk_e_destroy_dh(void);
 int uadk_e_bind_ecc(ENGINE *e);
 void uadk_e_destroy_ecc(void);
 int uadk_e_is_env_enabled(const char *alg_name);
+void uadk_e_set_env_enabled(const char *alg_name, unsigned int value);
```
Comment (CaptainMiaow) [inline src/uadk.h:43]: Doed the declaration of uadk_e_set_env_enabled() needs to be placed in uadk_utils.h file ?

**Comment by CaptainMiaow** [comment]
14:31:47 autoreconf: Leaving directory `.'
14:31:47 configure: error: cannot find sources (src/e_uadk.c) in . or ..

 The file name of e_uadk.c need changed to "uadk_engine_init.c"

File: src/uadk_digest.c [hunk 7]
```diff
@@ -927,3 +1017,153 @@ void uadk_e_destroy_digest(void)
...
+}
+
+/* Forward declaration of uadk implementation functions */
+static OSSL_FUNC_digest_freectx_fn	uadk_prov_freectx;
+static OSSL_FUNC_digest_dupctx_fn	uadk_prov_dupctx;
+static OSSL_FUNC_digest_init_fn		uadk_prov_init;
+static OSSL_FUNC_digest_update_fn	uadk_prov_update;
+static OSSL_FUNC_digest_final_fn	uadk_prov_final;
+static OSSL_FUNC_digest_gettable_params_fn
+					uadk_prov_gettable_params;
```
Comment (CaptainMiaow) [inline src/uadk_digest.c:1128]: The name "uadk_prov_gettable_params" maybe "uadk_prov_digest_gettable_params".

File: src/uadk_digest.c [hunk 8]
```diff
@@ -927,3 +1017,153 @@ void uadk_e_destroy_digest(void)
...
+	{ OSSL_FUNC_DIGEST_NEWCTX, (void (*)(void))uadk_##name##_newctx },	\
+	{ OSSL_FUNC_DIGEST_FREECTX, (void (*)(void))uadk_prov_freectx },	\
+	{ OSSL_FUNC_DIGEST_DUPCTX, (void (*)(void))uadk_prov_dupctx },		\
+	{ OSSL_FUNC_DIGEST_INIT, (void (*)(void))uadk_prov_init },		\
+	{ OSSL_FUNC_DIGEST_UPDATE, (void (*)(void))uadk_prov_update },		\
+	{ OSSL_FUNC_DIGEST_FINAL, (void (*)(void))uadk_prov_final },		\
+	{ OSSL_FUNC_DIGEST_GET_PARAMS,						\
+		(void (*)(void))uadk_##name##_get_params },			\
+	{ OSSL_FUNC_DIGEST_GETTABLE_PARAMS,					\
+		(void (*)(void))uadk_digest_gettable_params },			\
```
Comment (CaptainMiaow) [inline src/uadk_digest.c:1158]: The name "uadk_digest_gettable_params" maybe "uadk_prov_digest_gettable_params".

**Comment by docularxu** [comment]
Need to update README.md

**Comment by docularxu** [comment]
>  sha3-224 openssl speed -provider xx/uadk_provider.so -provider default -evp sha3-256 openssl speed -provider xx/uadk_provider.so -provider default -evp sha3-384 openssl speed -provider xx/uadk_provider.so -provider default -evp sha3-512

There is a mistake. What provided in this provider is SHA2, not SHA3.

 Provided:
  { 1.3.14.3.2.26, SHA1 } @ uadk_provider
  { 1.2.156.10197.1.401, SM3 } @ uadk_provider
  { 1.2.840.113549.2.5, MD5 } @ uadk_provider
  SHA2-224 @ uadk_provider
  SHA2-256 @ uadk_provider
  SHA2-384 @ uadk_provider
  SHA2-512 @ uadk_provider

please change  SHA3-224 to SHA2-224, ... like wise.

**Comment by gaozhangfei** [comment]
Update segmentation fault issue and Readme
also sanity_test.sh
    If ./configure --libdir=/usr/local/lib/ossl-modules/
    $ ./test/sanity_test.sh
    
    If ./configure --prefix=/tmp/build
    $ ./test/sanity_test.sh /tmp/build/lib/uadk_provider.so

**Pull request closed without merging**